### PR TITLE
[INLONG-6780][Sort] Supports dirty data side-output for JDBC connector and related sinks

### LINKED ISSUE #6780: [Feature][Sort] Supports dirty data side-output for JDBC connector and related sinks
### Description

Description
Supports dirty data side-output for jdbc related connectors.
In this part:

Load 'DirtySinkFactory' and create 'DirtySink' by the config
It is needs to determine whether it is dirty data in the connector.
Side output dirty data by the 'DirtySink' dependents on the configured, the built-in side-out of dirty data has 'LogDirtySink'(https://github.com/apache/inlong/pull/6618) and 'S3DirtySink'(https://github.com/apache/inlong/pull/6655).

### PR DESCRIPTION
**Pull request by Yizhou-Yang**
Fixes #6780 

## Motivation

Supports dirty data side-output for jdbc connector and all related fields.

In this part:
changed factories and builders to support dirty sink and options
adjusted DynamicTableSink to include new parameters
changed unit tests for mysql/sqlserver/oracle/pg to make it compatible with the changes (still working on it if it doesn't work now)

### CODE REVIEW

**Review by healchow — APPROVED** [review]

File: inlong-sort/sort-connectors/jdbc/src/main/java/org/apache/inlong/sort/jdbc/internal/JdbcBatchingOutputFormat.java [hunk 1]
```diff
@@ -147,6 +160,13 @@ public void open(int taskNumber, int numTasks) throws IOException {
         if (metricOption != null) {
             sinkMetricData = new SinkMetricData(metricOption, runtimeContext.getMetricGroup());
         }
+        if (dirtySink != null) {
+            try {
```
Comment (yunqingmoswu) [inline inlong-sort/sort-connectors/jdbc/src/main/java/org/apache/inlong/sort/jdbc/internal/JdbcBatchingOutputFormat.java:164]: Exceptions cannot be ignored during initialization.

File: inlong-sort/sort-connectors/jdbc/src/main/java/org/apache/inlong/sort/jdbc/internal/JdbcBatchingOutputFormat.java [hunk 2]
```diff
@@ -196,8 +216,39 @@ private void checkFlushException() {
...
+            throw ex;
+        }
+        if (dirtySink != null) {
+            DirtyData.Builder<Object> builder = DirtyData.builder();
+            try {
+                builder.setData(dirtyData)
+                        .setDirtyType(dirtyType)
+                        .setLabels(dirtyOptions.getLabels())
+                        .setLogTag(dirtyOptions.getLogTag())
+                        .setIdentifier(dirtyOptions.getIdentifier());
```
Comment (yunqingmoswu) [inline inlong-sort/sort-connectors/jdbc/src/main/java/org/apache/inlong/sort/jdbc/internal/JdbcBatchingOutputFormat.java:239]: Please add 'setDirtyMessage'.

File: inlong-sort/sort-connectors/jdbc/src/main/java/org/apache/inlong/sort/jdbc/internal/JdbcBatchingOutputFormat.java [hunk 3]
```diff
@@ -196,8 +216,39 @@ private void checkFlushException() {
         }
     }
 
+    void handleDirtyData(Object dirtyData, DirtyType dirtyType, Exception e) {
+        if (sinkMetricData != null) {
```
Comment (yunqingmoswu) [inline inlong-sort/sort-connectors/jdbc/src/main/java/org/apache/inlong/sort/jdbc/internal/JdbcBatchingOutputFormat.java:220]: Dirty metric reports should be location after '!dirtyOptions.ignoreDirty()'.

File: inlong-sort/sort-connectors/jdbc/src/main/java/org/apache/inlong/sort/jdbc/internal/JdbcBatchingOutputFormat.java [hunk 4]
```diff
@@ -214,11 +265,8 @@ public final synchronized void writeRecord(In record) throws IOException {
                 resetStateAfterFlush();
             }
         } catch (Exception e) {
-            if (sinkMetricData != null) {
-                sinkMetricData.invokeDirty(rowSize, dataSize);
-            }
-            resetStateAfterFlush();
-            throw new IOException("Writing records to JDBC failed.", e);
+            LOG.error(String.format("serialize error, raw data: %s", record), e);
+            handleDirtyData(record, DirtyType.SERIALIZE_ERROR, e);
```
Comment (yunqingmoswu) [inline inlong-sort/sort-connectors/jdbc/src/main/java/org/apache/inlong/sort/jdbc/internal/JdbcBatchingOutputFormat.java:269]: The type of dirty data is right?

**Review by gong — APPROVED** [review]

**Comment by gong** [comment]
no use merge, use git rebase master

**Review by EMsnap — APPROVED** [review]

**Review by yunqingmoswu — APPROVED** [review]

**Pull request merged**